### ISSUE TITLE
Strange focus problem

### ISSUE THREAD

**Issue by daliborp1976** [Komodo/KomodoEdit#903]
Komodo Edit, version 9.3.1, build 16424, platform win32-x86.
Built on Tue Nov 17 19:36:48 2015.

![komodo_focus_problem](https://cloud.githubusercontent.com/assets/16021097/11407873/db45eeba-93b5-11e5-80f0-76a84f10037d.png)

See atached CS.
I have several macros in toolbox, all of them doing some funky regex replacements. I execute them one after another, two of them most of the time. With this new Komodo release (see up) something strange happens. Double clicking on first macro (eg. BIB_RBR) - all is fine, macro fires and does the job. Now, i double click another macro (eg. BIB_P_REGULAR) and what happens is - previous macro fires (BIB_RBR)!
Shortly investigating I saw this; after first macro fires, focus on toolbox is lost, don't know where is it now(?). Next double clicking on another macro brings focus back to toolbox and previous macro fires. Which is, of course, wrong :)
But all works if I do thinng this way; double click first macro, all fine, focus lost. TWICE perform SINGLE click (not double) on desired second macro and it gets focus(!). Now do normal double click on second/desired macro and all is fine, it executes.
Except this proces is wrong :)

In short, executing two macros in a row actually results in two executions of a first macro.

**Comment by daliborp1976** [comment]
Err, any comment on this?
When working on large text/code projects and depending on macros - this bug is nearly a showstopper for using this version of Komodo :(

**Comment by Naatan** [comment]
Is the issue specific to the macro that you fire or does it happen on any tool?

Btw you can use Commando (Ctrl+Shift+O) to fire the macro as well.

**Comment by daliborp1976** [comment]
It happens with every tool or combination of different tools. Sometimes erratic but with macros all of the time exactly as I described.
To reporoduce, create two macros of any kind, put them in tools and execute as I described in original post.
Commando (go to anything) is not suitable for numerous repeatable sequence of macro executing.
I use macros to format numerous paragraphs of text or code.

**Comment by daliborp1976** [comment]
I just downgraded to 9.2.1 because this bug in combination with bug #902 is preventing me working on my project. Rather than switching to other tool I decided to downgrade and wait for fixes.
If I can help you with something about this issue, please let me know.

**Comment by Naatan** [comment]
I cannot reproduce this on Linux, @cgchoffman can you check on Windows?

**Comment by cgchoffman** [comment]
I think we need a couple sample Userscripts @daliborp1976.  I made two simple Userscripts that type different text to the buffer.  I double clicked back and forth between them as quickly as I could and I get the desired effect of "pizzabeerpizzabeerpizzabeer".

**Comment by daliborp1976** [comment]
Ok,
here it goes:

SAMPLE TEXT
<pre>
--------------------------------
(1)Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
incididunt ut labore et dolore magna aliqua. (2)Ut enim ad minim veniam, quis nostrud
exercitation ullamco laboris nisi ut (3)aliquip ex ea commodo consequat.
Duis aute irure dolor in reprehenderit in voluptate velit esse (4)cillum dolore eu
fugiat nulla pariatur. Excepteur sint occaecat (5)cupidatat non proident, sunt in culpa
qui officia deserunt mollit anim id est laborum.
---------------------------------
</pre>

MACRO1 (BIB_RBR)
<pre>
---------------------------------
// Macro recorded on: Wed Jun 24 2015 18:10:44 GMT+0200 (Central European Standard Time)
komodo.assertMacroVersion(3);
if (komodo.view) { komodo.view.setFocus(); }
ko.find.replaceAllInMacro(window, 1, '\\((\\d+)\\)\\s*', '&lt;span class=\"rbr\"&gt;(\\1)&lt;/span&gt;', false, 2, 0, false, false);
---------------------------------
</pre>

MACRO2 (BIB_P_REGULAR)
<pre>
---------------------------------
// Macro recorded on: Wed Jun 24 2015 18:43:20 GMT+0200 (Central European Standard Time)
komodo.assertMacroVersion(3);
if (komodo.view) { komodo.view.setFocus(); }
ko.find.replaceAllInMacro(window, 1, '((^.*?\\r\\n)*.+)', '&lt;p&gt;\\n\\1\\n&lt;/p&gt;\\n', false, 2, 0, false, false);
---------------------------------
</pre>

Pre-steps:
<pre>
create two attached macros
put them in toolbox
paste sample text in new/clean buffer
</pre>

Steps to reproduce bug:
<pre>
select all text (CTRL+A)
double click BIB_RBR macro - execute it
DO NOTING NOW, just double click BIB_P_REGULAR - execute it
now observe - what happend is BIB_RBR fired!
</pre>

Steps to do things right (but logically broken, fix needed):
<pre>
double click BIB_RBR macro - execute it (focus from toolbox is gone, somewhere)
SINGLE click BIB_P_REGULAR - nothing visible happens
SINGLE click BIB_P_REGULAR - macro name is now selected/hilighted
DOUBLE click BIB_P_REGULAR - execute it
ALL OK!
</pre>

**Comment by Naatan** [comment]
There's your problem:

`if (komodo.view) { komodo.view.setFocus(); }`

You are yourself shifting focus. You shouldn't need to do this for the code that you're running. Just remove those calls and you should be fine.

Also, for what it's worth, here is your first macro implemented with the new SDK:

```
var editor = require("ko/editor");
editor.replaceSelection(
    editor.getSelection().replace(/\((\d+)\)\s*/g, '<span class="rbr">($1)</span>')
);
```

**Comment by daliborp1976** [comment]
I did not write those macros by hand; they are recorded by Komodo itself! :)
Both in 9.2.1 and in 9.3.1.
Anyway, thank you for your help but you still have a problem because macro recorder produces non-optimal macros :)

**Comment by Naatan** [comment]
Oh really? Well then, I stand corrected.

**Comment by cgchoffman** [comment]
hi @daliborp1976, thank you for that info.  That really helped.  

> Anyway, thank you for your help but you still have a problem because macro recorder produces non-optimal macros :)

Komodo is actually doing exactly what you told it to do.  You likely clicked on the current view then activated find replace, am I correct?  I did the same and got the same result.  Try recording the same macro WITHOUT clicking on the editor view before you start the find replace.  I did this and it worked fine.

I don't think there is a bug here.

**Comment by daliborp1976** [comment]
I *might* agree but then there is a LOT to take into consideration when doing simple macro recording just to not get into weird *UI* response latter when executing macro :)
Don't you think? 
Why would macro steal focus from toolbox (it did not in 9.2.1) and why would toolbox then get confused where I click inside it?

Anyway - it's quite cumbersome all the way :)

**Comment by Naatan** [comment]
@daliborp1976 do you have to click twice always or is it possible only when a notification was showing?

Ie. if you click on the editor, and then back on the toolbox, does the first click just change focus?

For me this doesn't happen, unless a notification is showing.

**Comment by daliborp1976** [comment]
@Naatan you are right!
If I execute second macro while notification (from first macro action) is shown - described bug happens.
BUT! If I wait (do absolutely nothing) for notification to go away and then execute second macro - all is well!

So "hud" notification (while it's shown) is doing something weird and provokes described (wrong) UI response!

Good catch! I did not suspect notification to be the cause of this problem.

**Comment by Naatan** [comment]
Thanks @daliborp1976, we're tracking this here: #816

As a workaround you could disable certain types of notifications for now (hit the cog icon on the notification when it appears).
